libs/community/langchain_community/vectorstores/aperturedb.py

```diff
@@ -0,0 +1,446 @@
+        dimensions: Optional[int] = None,
+        engine: Optional[str] = None,
+        metric: Optional[str] = None,
+        log_level: int = logging.WARN,
+        properties: Optional[Dict] = None,
+        **kwargs: Any,
+    ) -> None:
+        """Create a vectorstore backed by ApertureDB
+
+        A single ApertureDB instance can support many vectorstores,
+        distinguished by 'descriptor_set' name.  The descriptor set is created
+        if it does not exist.  Different descriptor sets can use different
+        engines and metrics, be supplied by different embedding models, and have
+        different dimensions.
+
+        See ApertureDB documentation on `AddDescriptorSet`
+        https://docs.aperturedata.io/query_language/Reference/descriptor_commands/desc_set_commands/AddDescriptorSet
+        for more information on the engine and metric options.
+
+        Args:
+            embeddings (Embeddings): Embeddings object
+            descriptor_set (str, optional): Descriptor set name. Defaults to
+                "langchain".
+            dimensions (Optional[int], optional): Number of dimensions of the
+                embeddings. Defaults to None.
+            engine (str, optional): Engine to use. Defaults to "HNSW" for new
+                descriptorsets.
+            metric (str, optional): Metric to use. Defaults to "CS" for new
+                descriptorsets.
+            log_level (int, optional): Logging level. Defaults to logging.WARN.
+        """
+        # ApertureDB imports
+        try:
+            from aperturedb.Utils import Utils, create_connector
+        except ImportError:
+            raise ImportError(
+                "ApertureDB is not installed. Please install it using "
+                "'pip install aperturedb'"
+            )
+
+        super().__init__(**kwargs)
+        self.logger = logging.getLogger(__name__)
+        self.logger.setLevel(log_level)
+        self.descriptor_set = descriptor_set
+
+        self.embedding_function = embeddings
+        self.dimensions = dimensions
+        self.engine = engine
+        self.metric = metric
+        self.properties = properties
+        if embeddings is None:
+            self.logger.fatal("No embedding function provided.")
+            raise ValueError("No embedding function provided.")
+
+        try:
+            from aperturedb.Utils import Utils, create_connector
+        except ImportError:
+            self.logger.exception(
+                "ApertureDB is not installed. Please install it using "
+                "'pip install aperturedb'"
+            )
+            raise
+
+        self.connection = create_connector()
+        self.utils = Utils(self.connection)
+        try:
+            self.utils.status()
+        except Exception:
+            self.logger.exception("Failed to connect to ApertureDB")
+            raise
+
+        self._find_or_add_descriptor_set()
+
+    def _find_or_add_descriptor_set(self) -> None:
```

Comment:
FYI This makes network connections on init -- it'll complicate async support in the future.

Given that a lot of vectorstores currently do this, we'll accept the code as is

Reply:
Hmm.  We do prefer to create connections on init.  What is the async use case that will break with this?  Would it be better to open the connection lazily or do we need to be using multiple connections?

Reply:
Sorry thought I responded already! 

The issues are: 
1. you can't make an async call from init without kicking off a thread with a new event loop in it
2. users may also end up with a different life cycle for the vectorstore (e.g., instantiating it on every request served by a web server) -- this is not an async specific issue

Regardless it's fine for now, most vector stores implement this pattern